bootstrap/src/q.ml

```diff
@@ -0,0 +1,202 @@
+open Rudiments
+
+type 'a t = usize * 'a Stream.t * 'a list * 'a Stream.t
+
+let empty = (0, lazy Stream.Nil, [], lazy Stream.Nil)
+
+let exec (l, f, r, s) =
+  let rec rotate (l, f, r, s) = lazy begin
+    match f, r with
+    | _, [] -> assert false
+    | lazy Stream.Nil, y :: _ -> Lazy.force (Stream.push y s)
+    | lazy (Stream.Cons(elm, f')), y :: y' -> begin
+      let s' = Stream.push y s in
+      let f'' = rotate (l, f', y', s') in
+      Lazy.force (Stream.push elm f'')
+    end
+  end in
+  match s with
+  | lazy (Stream.Cons(_, s')) -> (l, f, r, s')
+  | lazy Stream.Nil -> begin
+    let f' = rotate (l, f, r, s) in
+    (l, f', [], f')
+  end
+
+let length (l, _, _, _) = l
+
+let is_empty (l, _, _, _) =
+  l = 0
+
+let hd (_, f, _, _) =
+  match f with
+  | lazy Stream.Nil -> halt "Empty q has no head"
+  | lazy (Stream.Cons(elm, _)) -> elm
+
+let tl (l, f, r, s) =
+  match f with
+  | lazy Stream.Nil -> halt "Empty q has no tail"
+  | lazy (Stream.Cons(_, f')) -> exec (pred l, f', r, s)
+
+let push_back (l, f, r, s) elm =
+  exec (succ l, f, elm :: r, s)
+
+let pop t =
+  hd t, tl t
+
+let pp pp_elm ppf (l, f, r, s) =
+  let open Format in
+  fprintf ppf "@[<h>";
+  fprintf ppf "(len=%a, f=%a, r=%a, s=%a)"
```

Comment:
The spaces in the format specifier should have `@` prefixes so that the Format machinery knows they are acceptable places to insert line breaks.  This isn't necessary in test code, but we should try to keep the `pp` functions capable of automatic wrapping for use in non-trivial formatted output cases.

  `"(len=%a,@ f=%a,@ r=%a,@ s=%a)"`